bentoml/_internal/exportable.py

```diff
@@ -0,0 +1,363 @@
+import os
+import re
+import typing as t
+import urllib.parse
+from abc import ABC
+from abc import abstractmethod
+import logging
+
+import fs
+import fs.copy
+import fs.mirror
+import fs.opener
+from fs.base import FS
+
+T = t.TypeVar("T")
+
+
+uriSchemeRe = re.compile(r".*[^\\](?=://)")
+
+
+class Exportable(ABC):
+    @staticmethod
+    @abstractmethod
+    def _export_ext() -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def _export_name(self) -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def _fs(self) -> FS:
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def from_fs(cls: t.Type[T], item_fs: FS) -> T:
+        raise NotImplementedError
+
+    @classmethod
+    def guess_format(cls, path: str) -> str:
+        _, ext = fs.path.splitext(path)
+
+        ext = ext if ext == "" else ext[1:]
+        if ext in [cls._export_ext(), "gz", "xz", "bz2", "tar", "zip"]:
+            return ext
+        else:
+            return cls._export_ext()
+
+    @classmethod
+    def import_from(
+        cls: t.Type[T],
+        path: str,
+        input_format: t.Optional[str] = None,
```

Comment:
what do you think about changing this type to a `t.Union[None, t.Literal[ SUPPORT_FORMAT_LIST ]]`? which would serve as a good documentation for supported formats, and similarly for the `protocol` below

Reply:
nvm, I think documenting it at the top-level user facing API is good enough